[RISCV] Enable TTI::shouldDropLSRSolutionIfLessProfitable by default

asb opened the pull request:
This avoids some cases where LSR produces results that lead to very poor codegen. There's a chance we'll see minor degradations for some inputs in the case that our metrics say the found solution is worse, but in reality it's better than the starting point.

---

Stacks on https://github.com/llvm/llvm-project/pull/89924

Not tested enough yet to realistically merge at this point, but posting early in case others have looked at this already. Here is an example input that has terrible codegen without this:

```llvm
%struct.ham = type { i64, i32, i32, i32, i32, i32, i32, i32, i32, i64, i32, i64, i64, i32, i64 }

define i32 @main() {
bb:
  %call = tail call ptr null(i64 0)
  br label %bb2

bb1:                                              ; No predecessors!
  %load = load i32, ptr %call, align 4
  ret i32 0

bb2:                                              ; preds = %bb2, %bb
  %phi = phi i64 [ 0, %bb ], [ %add, %bb2 ]
  %getelementptr = getelementptr %struct.ham, ptr %call, i64 %phi
  %getelementptr3 = getelementptr i8, ptr %getelementptr, i64 8
  store i32 0, ptr %getelementptr3, align 8
  %getelementptr4 = getelementptr i8, ptr %getelementptr, i64 12
  store i32 0, ptr %getelementptr4, align 4
  %getelementptr5 = getelementptr i8, ptr %getelementptr, i64 16
  store i32 0, ptr %getelementptr5, align 8
  %getelementptr6 = getelementptr i8, ptr %getelementptr, i64 20
  store i32 0, ptr %getelementptr6, align 4
  %getelementptr7 = getelementptr i8, ptr %getelementptr, i64 24
  store i32 0, ptr %getelementptr7, align 8
  %getelementptr8 = getelementptr i8, ptr %getelementptr, i64 28
  store i32 0, ptr %getelementptr8, align 4
  %getelementptr9 = getelementptr i8, ptr %getelementptr, i64 32
  store i32 0, ptr %getelementptr9, align 8
  %getelementptr10 = getelementptr i8, ptr %getelementptr, i64 36
  store i32 0, ptr %getelementptr10, align 4
  %getelementptr11 = getelementptr i8, ptr %getelementptr, i64 40
  store i64 0, ptr %getelementptr11, align 8
  %getelementptr12 = getelementptr i8, ptr %getelementptr, i64 48
  store i32 0, ptr %getelementptr12, align 8
  %getelementptr13 = getelementptr i8, ptr %getelementptr, i64 72
  store i32 0, ptr %getelementptr13, align 8
  %getelementptr14 = getelementptr i8, ptr %getelementptr, i64 80
  store i64 0, ptr %getelementptr14, align 8
  %add = add i64 %phi, 1
  br label %bb2
}
```

There's probably an LSR change related to its search heuristic that would address this (drop the last gep+store and LSR finds a good solution), but if LSR finds something that by it's own metrics is much worse than the starting point then going ahead with it seems hard to justify.

asb commented:
Although blocked on #89924 moving the option to a hook, I'd be interested if anyone has any feedback or thoughts on this direction. Dropping LSR solutions that our metrics deem unprofitable, and working to improve the heuristics if that causes any unexpected degradation seems sensible to me. But there may be history with work in this area I'm not aware of.

preames commented on llvm/test/CodeGen/RISCV/rvv/dont-sink-splat-operands.ll:
```diff
@@ -86,30 +86,29 @@ declare i64 @llvm.vscale.i64()
+; NO-SINK-NEXT:    add a6, a0, a6
 ; NO-SINK-NEXT:    vl2re32.v v10, (a6)
 ; NO-SINK-NEXT:    vadd.vv v10, v10, v8
+; NO-SINK-NEXT:    add a5, a5, a2
```
These diffs seem likely unprofitable, but more than that, they seems *weird*.  The solution being chosen should "clearly" be higher cost than the one being rejected here.  But more than that, the previous solution was also sub-optimal.  

At first, I thought this might relate to the shifted add and went looking for a case where we forgot to guard a shNadd instruction on Zba.  However, that turned out to be a red herring.

However, when doing that investigation, I noticed something.  We have *two* different implementations of isLegalAddressing mode.  One on TLI and one on TTI.  We have overridden the former, but don't appear to have done the later.

The default implementation on TTIImpl assumes reg+reg addressing is legal.  For RISCV vector, this is not a legal addressing mode.  I think what we're seeing here is that LSR is computing a bad cost for basically all addresses used by vector loads and stores, and that the actual effect of this change is basically noise.  

I suspect that if we implement TTI::isLegalAddressing, we'll see significant improvements in LSR solutions.  Once we do that, we can revisit this patch.

preames replied:
It looks like I got myself confused here.  BasicTTI does perform the proxy to TLI, so these implementations should be the same.

asb commented:
Thank you for taking a look. I've gone ahead and rebased to check that the changes remain similar.

My main goal of posting this quickly was to see if there was any memory on previous efforts in this direction given the original LSR flag came from RISC-V contributors. It sounds like there's not!

I agree that the changes to the in-tree vector code are a minor degradation - multiple cases where we're we have an additional scalar instruction in the loop. Not a big regression, but it's something and probably worth understanding. This patch also misses the test/Transforms/LoopStrengthReduce/RISCV changes that merit a closer look.

"Noise" is a good description and roughly what I would expect outside of the case I found or those like it.  The heuristics for LSR seem very rough and potentially misleading in some cases, so although I think dropping solutions we view as less profitable than the starting point is a sensible way to go and a better basis on which to build any further LSR improvements, it's likely to cause minor variations that aren't much better or worse in many cases.

I think based on the discussion here the next sensible steps are:
* Precommit the test case in the PR
  * It seems a useful example to have coverage for regardless of whether this ends up landing or not.
* Look more closely at why we see a minor regression in some of the changed vector test cases
* Run across a wider range of external inputs to look for surprises
  * I'd wondered if people had already done this, but it seems not

topperc commented:
We've been using -lsr-drop-solution for some benchmarks in our downstream for a while. Overall, I think it has been a net improvement.

preames commented:
> We've been using -lsr-drop-solution for some benchmarks in our downstream for a while. Overall, I think it has been a net improvement.

This is good to hear.  I have no downstream experience with this change, but also have no objection to landing this.

Given that, LGTM.  We can iterate to improve the couple of regressions in tree.

asb commented:
Sorry for the delay, I was away on holiday. I've rebased, precommitted the example I used in the original description, and updated the affected tests in llvm/test/Transforms/LoopStrengthReduce/RISCV that were missed in the first patch.

@preames I think you wrote the affected LoopStrengthReduce/RISCV tests - I just wanted to reconfirm you're happy with the test changes shown in these files, and if you think any additional comments should be added.

preames commented:
> @preames I think you wrote the affected LoopStrengthReduce/RISCV tests - I just wanted to reconfirm you're happy with the test changes shown in these files, and if you think any additional comments should be added.

The reasoning from my early LGTM still applies.  I glanced at the LSR changes, and see a couple of regressions, but I still think the right path forward is to enable this and then iterative improve.

topperc commented:
What's the status of this? @asb can you rebase?

asb committed:
[RISCV] Enable TTI::shouldDropLSRSolutionIfLessProfitable by default

This avoids some cases where LSR produces results that lead to very poor
codegen. There's a chance we'll see minor degradations for some inputs
in the case that our metrics say the found solution is worse, but in
reality it's better than the starting point.

Per the review thread, at least one vendor has been enabling this by
defualt for some time and found overall it's an improvement. As such,
we'll enable by default and aim to fix any as-yet-unknown regressions
in-tree.

asb commented:
> What's the status of this? @asb can you rebase?

I think given the positive review from Philip and your note that you've been using this downstream for some time it's good to land. I'd missed that I hadn't actually hit the merge button yet. I've rebased and will merge now.

dtcxzyw commented:
Hi @asb, I found this patch caused some significant performance regression (up to 37.8%) on llvm-test-suite with `rv64gczba_zbb_zbkb_zbs -O3 -flto=thin`. Can you have a look about this?

| Name                                                         | Before (Billion Insts) | After (Billion Insts) | Ratio |
| ------------------------------------------------------------ | ---------------------- | --------------------- | ----- |
| SingleSource/Benchmarks/Polybench/linear-algebra/kernels/doitgen/doitgen | 4.222953611            | 5.820735229           | 1.378 |
| SingleSource/Benchmarks/Polybench/datamining/covariance/covariance | 4.090009899            | 5.593993395           | 1.368 |
| SingleSource/Benchmarks/Polybench/datamining/correlation/correlation | 4.095016076            | 5.599001576           | 1.367 |
| SingleSource/Benchmarks/Polybench/medley/reg_detect/reg_detect | 0.206776896            | 0.267266747           | 1.293 |
| SingleSource/Benchmarks/Shootout/Shootout-ary3               | 10.507102477           | 13.507101469          | 1.286 |
| SingleSource/Benchmarks/Polybench/linear-algebra/kernels/mvt/mvt | 0.576970074            | 0.720934057           | 1.250 |
| SingleSource/Benchmarks/Polybench/linear-algebra/solvers/durbin/durbin | 0.535427699            | 0.663343689           | 1.239 |
| SingleSource/Benchmarks/Polybench/linear-algebra/kernels/trisolv/trisolv | 0.160518143            | 0.192501995           | 1.199 |
| SingleSource/Benchmarks/Polybench/stencils/adi/adi           | 3.382377197            | 4.017469438           | 1.188 |
| MultiSource/Benchmarks/MiBench/automotive-susan/automotive-susan | 0.284736344            | 0.334505345           | 1.175 |
| SingleSource/Benchmarks/Polybench/linear-algebra/kernels/atax/atax | 0.368520477            | 0.43248861            | 1.174 |
| SingleSource/Benchmarks/BenchmarkGame/puzzle                 | 3.092676262            | 3.592676238           | 1.162 |
| SingleSource/Benchmarks/Misc/himenobmtxpa                    | 2.368697352            | 2.746954721           | 1.160 |
| MultiSource/Benchmarks/Rodinia/srad/srad                     | 2.247984884            | 2.600044498           | 1.157 |

For detailed performance data, please refer to https://github.com/dtcxzyw/llvm-ci/issues/1165 and https://github.com/dtcxzyw/llvm-ci/pull/1166.

asb commented:
@dtcxzyw thank you, that's definitely more impact than I was expecting so I've reverted while it's still easy to do so in order to allow proper investigation.
